src/neo/SmartContract/Native/NameService.cs

```diff
@@ -0,0 +1,265 @@
+#pragma warning disable IDE0051
+
+using Neo.Cryptography;
+using Neo.IO;
+using Neo.IO.Json;
+using Neo.Ledger;
+using Neo.Persistence;
+using Neo.VM;
+using Neo.VM.Types;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Numerics;
+using System.Text.RegularExpressions;
+
+namespace Neo.SmartContract.Native
+{
+    public sealed class NameService : NonfungibleToken<NameService.NameState>
+    {
+        public override int Id => -6;
+        public override string Symbol => "NNS";
+
+        private const byte Prefix_Roots = 10;
+        private const byte Prefix_DomainPrice = 22;
+        private const byte Prefix_Expiration = 20;
+        private const byte Prefix_Record = 12;
+
+        private const uint OneYear = 365 * 24 * 3600;
+        private static readonly Regex rootRegex = new Regex("^[a-z][a-z0-9]{0,15}$", RegexOptions.Singleline);
+        private static readonly Regex nameRegex = new Regex("^(?=.{3,255}$)([a-z0-9]{1,62}\\.)+[a-z][a-z0-9]{0,15}$", RegexOptions.Singleline);
```

Comment:
We should not allow '\n' at the end or some characters that can facilitate phishing

Reply:
Please check https://github.com/neo-project/neo/pull/2201/files#diff-35ad071f1b3c3545b382dd014b359a1aab6821989d280fc3e1c10475c2735cfaR105